src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,375 @@
+
+    Returns:
+        `torch.nn.Module`: The X-LoRA model.
+
+    Example:
+        ```py
+        >>> from transformers import AutoModelForCausalLM, AutoConfig
+        >>> from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_int8_training
+
+        >>> model_config = AutoConfig.from_pretrained("mistralai/Mistral-7B-Instruct-v0.1")
+        >>> config = XLoraConfig(
+        ...     task_type="CAUSAL_LM",
+        ...     hidden_size=model_config.hidden_size,
+        ...     xlora_depth=4,
+        ...     adapters={
+        ...         "adapter_1": "./path/to/the/checkpoint/",
+        ...         "adapter_2": "./path/to/the/checkpoint/",
+        ...         "adapter_n": "./path/to/the/checkpoint/",
+        ...     },
+        ... )
+
+        >>> model = AutoModelForCausalLM.from_pretrained(
+        ...     "mistralai/Mistral-7B-Instruct-v0.1",
+        ...     trust_remote_code=True,
+        ...     use_flash_attention_2=False,
+        ...     device_map="cuda:0",
+        ...     torch_dtype=torch.bfloat16,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> xlora_model = get_peft_model(model, config)
+        ```
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        config: Union[dict[str, XLoraConfig], XLoraConfig],
+        adapter_name: str,
+    ) -> None:
+        nn.Module.__init__(self)
+
+        if isinstance(config, dict):
+            conf = config[adapter_name]
+        else:
+            conf = config
+        lora_model = LoraModel(model, config.copy(), adapter_name)
+        self.xlora_config = conf
+        del self.xlora_config.target_modules
+        self.lora_model = lora_model
+
+        peft_config = conf
+
+        if hasattr(model.config, "use_cache") and model.config.use_cache:
+            raise ValueError("`use_cache` must be False")
+
+        adapters_items = peft_config.adapters.items()
+        if hasattr(self.xlora_config, "_subfolders"):
+            adapters_items = zip(peft_config.adapters.items(), self.xlora_config._subfolders)
+        else:
+            adapters_items = peft_config.adapters.items()
+
+        if hasattr(self.xlora_config, "_subfolders"):
+            for (adapter_name, model_id), subfolder in adapters_items:
+                self.lora_model.load_adapter(model_id, adapter_name, subfolder=subfolder)
+        else:
+            for adapter_name, model_id in adapters_items:
+                self.lora_model.load_adapter(model_id, adapter_name)
+
+        self.lora_model.set_adapter(list(peft_config.adapters.keys()))
+
+        self._maybe_freeze_all_adapters()
+
+        total_swapped, device, all_layers = convert_layers_to_xlora(
+            model,
+            self,
+            peft_config,
+        )
+
+        # Now replace the old forward function with a new one that implements the X-LoRA architecture
+        old_model_forward = self.lora_model.model.forward
```

Comment:
Here we still rely on replacing the forward method and passing the scaling parameter by setting it as an attribute on the layer. Did you consider if this could be made more straightforward with the forward hooks added via `_enable_peft_forward_hooks` as I suggested in my PR?